File: src/lib/backup/backupChallenge.test.ts
import { describe, expect, it } from "vitest";

import {
  createBackupChallenge,
  isBackupAnswerCorrect,
  isBackupChallengeComplete,
} from "./backupChallenge";

const FRASE =
  "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about";

describe("sorteio das posições", () => {
  it("sorteia 3 posições distintas, em ordem", () => {
    for (let i = 0; i < 200; i++) {
      const c = createBackupChallenge(12);
      expect(c.positions).toHaveLength(3);
      expect(new Set(c.positions).size).toBe(3);
      expect([...c.positions].sort((a, b) => a - b)).toEqual([...c.positions]);
      expect(c.positions.every((p) => p >= 0 && p < 12)).toBe(true);
    }
  });

  it("não sorteia mais perguntas do que palavras", () => {
    expect(() => createBackupChallenge(2)).toThrow();
  });
});

describe("conferência das respostas", () => {
  it("aceita a palavra certa, com espaço e maiúscula", () => {
    expect(isBackupAnswerCorrect(FRASE, 11, "about")).toBe(true);
    expect(isBackupAnswerCorrect(FRASE, 11, "  ABOUT ")).toBe(true);
  });

  it("recusa palavra errada e posição fora da frase", () => {
    expect(isBackupAnswerCorrect(FRASE, 11, "abandon")).toBe(false);
    expect(isBackupAnswerCorrect(FRASE, 99, "about")).toBe(false);
    expect(isBackupAnswerCorrect(FRASE, -1, "abandon")).toBe(false);
  });

  it("só conclui quando TODAS as sorteadas estão certas", () => {
    const c = { positions: [0, 5, 11] as const };
    expect(
      isBackupChallengeComplete(
        FRASE,
        c,
        new Map([[0, "abandon"], [5, "abandon"], [11, "about"]]),
      ),
    ).toBe(true);
    expect(
      isBackupChallengeComplete(
        FRASE,
        c,
        new Map([[0, "abandon"], [5, "abandon"], [11, "abandon"]]),
      ),
    ).toBe(false);
    expect(isBackupChallengeComplete(FRASE, c, new Map([[0, "abandon"]]))).toBe(false);
  });
});

File: test/backup_challenge_test.dart
import 'dart:math';

import 'package:bitibridge_wallet/src/backup/backup_challenge.dart';
import 'package:flutter_test/flutter_test.dart';

const String kFrase =
    'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about';

void main() {
  group('sorteio das posições', () {
    test('sorteia 3 posições distintas, em ordem', () {
      for (int i = 0; i < 200; i++) {
        final BackupChallenge c = createBackupChallenge(12);
        expect(c.positions.length, 3);
        expect(c.positions.toSet().length, 3, reason: 'não pode repetir palavra');
        final List<int> ordenado = List<int>.from(c.positions)..sort();
        expect(c.positions, ordenado);
        expect(c.positions.every((int p) => p >= 0 && p < 12), isTrue);
      }
    });

    test('numeração mostrada ao usuário começa em 1', () {
      final BackupChallenge c =
          createBackupChallenge(12, rng: Random(42));
      expect(c.humanPositions, c.positions.map((int p) => p + 1).toList());
      expect(c.humanPositions.every((int p) => p >= 1 && p <= 12), isTrue);
    });

    test('não sorteia mais perguntas do que palavras', () {
      expect(() => createBackupChallenge(2), throwsArgumentError);
    });
  });

  group('conferência das respostas', () {
    test('aceita a palavra certa, com espaço e maiúscula', () {
      expect(isBackupAnswerCorrect(kFrase, 11, 'about'), isTrue);
      expect(isBackupAnswerCorrect(kFrase, 11, '  ABOUT '), isTrue);
      expect(isBackupAnswerCorrect(kFrase, 0, 'abandon'), isTrue);
    });

    test('recusa palavra errada e posição fora da frase', () {
      expect(isBackupAnswerCorrect(kFrase, 11, 'abandon'), isFalse);
      expect(isBackupAnswerCorrect(kFrase, 99, 'about'), isFalse);
      expect(isBackupAnswerCorrect(kFrase, -1, 'abandon'), isFalse);
    });

    test('só conclui quando TODAS as sorteadas estão certas', () {
      const BackupChallenge c = BackupChallenge(<int>[0, 5, 11]);
      expect(
        isBackupChallengeComplete(kFrase, c, <int, String>{
          0: 'abandon',
          5: 'abandon',
          11: 'about',
        }),
        isTrue,
      );
      // uma errada derruba tudo
      expect(
        isBackupChallengeComplete(kFrase, c, <int, String>{
          0: 'abandon',
          5: 'abandon',
          11: 'abandon',
        }),
        isFalse,
      );
      // faltando resposta também
      expect(
        isBackupChallengeComplete(kFrase, c, <int, String>{0: 'abandon'}),
        isFalse,
      );
    });
  });
}
